pkg/storage-init/storage-init.sh

```diff
@@ -34,6 +34,57 @@ fi
 
 P3_FS_TYPE="ext3"
 FSCK_FAILED=0
+
+# First lets see if we're missing P3 altogether and try to create it.
+# This is safe, since the worst case scenario we may hose the system,
+# but the system without P3 is a warm brick anyways
+P3=$(/hostfs/sbin/findfs PARTLABEL=P3)
+IMGA=$(/hostfs/sbin/findfs PARTLABEL=IMGA)
+if [ -z "$P3" ] && [ -n "$IMGA" ]; then
+   DEV=$(echo /sys/block/*/"${IMGA#/dev/}")
+   DEV="/dev/$(echo "$DEV" | cut -f4 -d/)"
+
+   # if sgdisk complains we need to repair the GPT
+   if sgdisk -v "$DEV" | grep -q 'Identified.*problems'; then
+       # save a copy of the MBR + first partition entry
+       # the logic here is that whatever booted us was good
+       # enough to get us here, so we'd rather sgdisk disn't
+       # mess up a good thing for us
+       dd if="$DEV" of=/tmp/mbr.bin bs=1 count=$((446 + 16)) conv=noerror,sync,notrunc
+
+       sgdisk -h1 -e "$DEV"
+
+       # move 1st MBR entry to 2nd place
+       dd if="$DEV" of="$DEV" bs=1 skip=446 seek=$(( 446 + 16)) count=16 conv=noerror,sync,notrunc
+       # restore 1st MBR entry + first partition entry
+       dd if=/tmp/mbr.bin of="$DEV" bs=1 conv=noerror,sync,notrunc
+
+       # focrce kernel to re-scan partition table
+       partprobe "$DEV"
+   fi
+
+   # lets see if IMGB partition is around, if not - create it
+   IMGB=$(/hostfs/sbin/findfs PARTLABEL=IMGB)
+   if [ -z "$IMGB" ]; then
+      IMGA_ID=$(sgdisk -p "$DEV" | grep "IMGA$" | awk '{print $1;}')
+
+      IMGA_SIZE=$(sgdisk -i "$IMGA_ID" "$DEV" | awk '/^Partition size:/ { print $3; }')
+      IMGA_GUID=$(sgdisk -i "$IMGA_ID" "$DEV" | awk '/^Partition unique GUID:/ { print $4; }')
+
+      SEC_START=$(sgdisk -f "$DEV")
+      SEC_END=$((SEC_START + IMGA_SIZE))
+      IMGB_ID=$((IMGA_ID + 1))
+
+      sgdisk --new "$IMGB_ID:$SEC_START:$SEC_END" \
+             --typecode="$IMGB_ID:$IMGA_GUID" --change-name="$IMGB_ID:IMGB" "$DEV"
```

Comment:
Same thing here. This isn't really part of "if persist partition doesn't exist, create it," but rather its own function that says, "ensure IMGB partition exists". Let's move it up and out to its own thing. Then the "ensure persist partition exists" becomes super clean."

Reply:
The problem here is the case where P3 does exist but IMGB doesn't. Then it becomes tricky to actually figure out where to put IMGB on disk even. I'd rather keep it here for now and refactor later if you don't mind (simply because I'm not quite sure what a sensible strategy for missing IMGB would be -- bail?)